File: 1_카카오/프로그래머스Lv2-[3차] 파일명 정렬.js
// isNumber()는 이렇게 써야한다.
const isNumber = (s) => {
  return !Number.isNaN(parseInt(s, 10))
}

const convertFile = (file) => {
  const N = file.length
  let HEAD = ""
  let NUMBER = "0"
  let HEAD_OK = false
  let NUMBER_OK = false
  for (let i = 0; i < N; i += 1) {
    const s = file[i]
    if (HEAD_OK && NUMBER_OK) {
      break
    }
    if (!HEAD_OK) {
      HEAD += s
      if (i < N - 1 && isNumber(file[i + 1])) {
        HEAD_OK = true
      }
      continue
    }
    NUMBER += s
    if (i < N - 1 && !isNumber(file[i + 1])) {
      NUMBER_OK = true
    }
  }
  return [HEAD, NUMBER]
}

const compare = (a, b) => {
  const [aHEAD, aNUMBER, aIndex] = a
  const [bHEAD, bNUMBER, bIndex] = b
  if (aHEAD !== bHEAD) {
    return aHEAD < bHEAD ? -1 : 1
  }
  if (aNUMBER !== bNUMBER) {
    return aNUMBER - bNUMBER
  }
  return aIndex - bIndex
}

function solution(files) {
  return files
    .map((file, i) => {
      const [HEAD, STRING_NUMBER] = convertFile(file.toLowerCase())
      return [HEAD, Number(STRING_NUMBER), i]
    })
    .sort(compare)
    .map(([, , index]) => files[index])
}
// 정확성  테스트
// 테스트 1 〉	통과 (0.46ms, 33.4MB)
// 테스트 2 〉	통과 (0.30ms, 33.4MB)
// 테스트 3 〉	통과 (13.05ms, 37.8MB)
// 테스트 4 〉	통과 (8.06ms, 38.1MB)
// 테스트 5 〉	통과 (6.37ms, 37.9MB)
// 테스트 6 〉	통과 (6.44ms, 37.7MB)
// 테스트 7 〉	통과 (8.51ms, 37.7MB)
// 테스트 8 〉	통과 (6.66ms, 37.9MB)
// 테스트 9 〉	통과 (6.96ms, 37.7MB)
// 테스트 10 〉	통과 (7.26ms, 37.9MB)
// 테스트 11 〉	통과 (6.35ms, 38.9MB)
// 테스트 12 〉	통과 (7.38ms, 39.2MB)
// 테스트 13 〉	통과 (6.39ms, 37.4MB)
// 테스트 14 〉	통과 (4.76ms, 37.6MB)
// 테스트 15 〉	통과 (4.29ms, 37.7MB)
// 테스트 16 〉	통과 (9.70ms, 37.9MB)
// 테스트 17 〉	통과 (4.18ms, 37.5MB)
// 테스트 18 〉	통과 (5.84ms, 37.8MB)
// 테스트 19 〉	통과 (6.63ms, 39.1MB)
// 테스트 20 〉	통과 (7.94ms, 37.7MB)

// 정규표현식 match
const compare = (a, b) => {
  const [aHEAD, aNUMBER, aIndex] = a
  const [bHEAD, bNUMBER, bIndex] = b
  if (aHEAD !== bHEAD) {
    return aHEAD < bHEAD ? -1 : 1
  }
  if (aNUMBER !== bNUMBER) {
    return aNUMBER - bNUMBER
  }
  return aIndex - bIndex
}

function solution(files) {
  const regExp = /\D+(\d{1,5})/i

  return files
    .map((file, index) => {
      const [searched, STRING_NUMBER] = file.match(regExp)
      const HEAD = searched.replace(STRING_NUMBER, "").toLowerCase()
      return [HEAD, parseInt(STRING_NUMBER, 10), index]
    })
    .sort(compare)
    .map((x) => files[x[2]])
}
// 정확성  테스트
// 테스트 1 〉	통과 (1.72ms, 33.6MB)
// 테스트 2 〉	통과 (1.88ms, 33.6MB)
// 테스트 3 〉	통과 (13.23ms, 37.7MB)
// 테스트 4 〉	통과 (8.92ms, 37.9MB)
// 테스트 5 〉	통과 (10.77ms, 37.7MB)
// 테스트 6 〉	통과 (6.30ms, 37.8MB)
// 테스트 7 〉	통과 (6.50ms, 37.7MB)
// 테스트 8 〉	통과 (7.16ms, 37.7MB)
// 테스트 9 〉	통과 (10.36ms, 37.7MB)
// 테스트 10 〉	통과 (6.23ms, 37.8MB)
// 테스트 11 〉	통과 (6.01ms, 37.8MB)
// 테스트 12 〉	통과 (8.14ms, 37.8MB)
// 테스트 13 〉	통과 (4.19ms, 34.6MB)
// 테스트 14 〉	통과 (3.89ms, 34.6MB)
// 테스트 15 〉	통과 (3.12ms, 34.7MB)
// 테스트 16 〉	통과 (7.14ms, 37.8MB)
// 테스트 17 〉	통과 (5.68ms, 34.5MB)
// 테스트 18 〉	통과 (5.65ms, 37.9MB)
// 테스트 19 〉	통과 (7.45ms, 37.7MB)
// 테스트 20 〉	통과 (5.93ms, 37.7MB)

// 정규표현식 기준으로 문자열 split : 이런 것도 된다... ㄷㄷ
const compare = (a, b) => {
  const [aHEAD, aNUMBER, aIndex] = a
  const [bHEAD, bNUMBER, bIndex] = b
  if (aHEAD !== bHEAD) {
    return aHEAD < bHEAD ? -1 : 1
  }
  if (aNUMBER !== bNUMBER) {
    return aNUMBER - bNUMBER
  }
  return aIndex - bIndex
}

function solution(files) {
  const regExp = /^(\D+)(\d{1,5})/i

  return files
    .map((file, index) => {
      const [, HEAD, STRING_NUMBER] = file.split(regExp)
      return [HEAD.toLowerCase(), parseInt(STRING_NUMBER, 10), index]
    })
    .sort(compare)
    .map((x) => files[x[2]])
}

File: 1_카카오/프로그래머스Lv2-[3차] 파일명 정렬.py
def convert_file(file):
    N = len(file)
    HEAD = ""
    NUMBER = "0"
    HEAD_OK = NUMBER_OK = False
    for i, s in enumerate(file):
        if HEAD_OK and NUMBER_OK:
            break
        if not HEAD_OK:
            HEAD += s
            if i < N - 1 and file[i + 1].isdigit():
                HEAD_OK = True
            continue
        NUMBER += s
        if i < N - 1 and not file[i + 1].isdigit():
            NUMBER_OK = True
    return [HEAD, NUMBER]

def solution(files):
    converted = []
    for i, file in enumerate(files):
        HEAD, STRING_NUMBER = convert_file(file.lower())
        converted.append([HEAD, int(STRING_NUMBER), i])
    
    return [files[x[2]] for x in sorted(converted)]
"""
정확성  테스트
테스트 1 〉	통과 (0.03ms, 10.5MB)
테스트 2 〉	통과 (0.03ms, 10.4MB)
테스트 3 〉	통과 (2.77ms, 10.4MB)
테스트 4 〉	통과 (2.86ms, 10.5MB)
테스트 5 〉	통과 (2.70ms, 10.6MB)
테스트 6 〉	통과 (2.74ms, 10.4MB)
테스트 7 〉	통과 (2.78ms, 10.5MB)
테스트 8 〉	통과 (2.42ms, 10.4MB)
테스트 9 〉	통과 (2.53ms, 10.5MB)
테스트 10 〉	통과 (4.78ms, 10.5MB)
테스트 11 〉	통과 (4.55ms, 10.3MB)
테스트 12 〉	통과 (2.77ms, 10.5MB)
테스트 13 〉	통과 (2.33ms, 10.6MB)
테스트 14 〉	통과 (4.70ms, 10.6MB)
테스트 15 〉	통과 (3.11ms, 10.5MB)
테스트 16 〉	통과 (2.51ms, 10.4MB)
테스트 17 〉	통과 (2.13ms, 10.5MB)
테스트 18 〉	통과 (4.35ms, 10.5MB)
테스트 19 〉	통과 (2.54ms, 10.4MB)
테스트 20 〉	통과 (2.50ms, 10.6MB)
"""

# 정규식 사용
import re
def solution(files):
    regexp = re.compile(r"(^\D+)(\d{1,5})", re.I)
    
    converted = []
    for i, file in enumerate(files):
        HEAD, STRING_NUMBER = regexp.match(file).groups()
        converted.append([HEAD.lower(), int(STRING_NUMBER), i])
    
    return [files[x[2]] for x in sorted(converted)]
"""
정확성  테스트
테스트 1 〉	통과 (0.14ms, 10.2MB)
테스트 2 〉	통과 (0.17ms, 10.2MB)
테스트 3 〉	통과 (1.38ms, 10.5MB)
테스트 4 〉	통과 (2.46ms, 10.5MB)
테스트 5 〉	통과 (2.11ms, 10.4MB)
테스트 6 〉	통과 (1.51ms, 10.3MB)
테스트 7 〉	통과 (2.44ms, 10.4MB)
테스트 8 〉	통과 (1.88ms, 10.5MB)
테스트 9 〉	통과 (1.34ms, 10.4MB)
테스트 10 〉	통과 (2.26ms, 10.3MB)
테스트 11 〉	통과 (1.94ms, 10.5MB)
테스트 12 〉	통과 (1.66ms, 10.4MB)
테스트 13 〉	통과 (1.02ms, 10.4MB)
테스트 14 〉	통과 (2.97ms, 10.7MB)
테스트 15 〉	통과 (1.65ms, 10.6MB)
테스트 16 〉	통과 (1.32ms, 10.4MB)
테스트 17 〉	통과 (1.39ms, 10.5MB)
테스트 18 〉	통과 (1.26ms, 10.4MB)
테스트 19 〉	통과 (1.45ms, 10.3MB)
테스트 20 〉	통과 (2.10ms, 10.3MB)
"""